File: 0499-The Maze III/0499-The Maze III.cpp
class Solution {
public:
    string findShortestWay(vector<vector<int>>& maze, vector<int>& ball, vector<int>& hole) {
        int m = maze.size();
        int n = maze[0].size();
        priority_queue<Cell> Q;
        vector<vector<int>> distance(m, vector<int>(n, INT_MAX));
        Q.emplace(ball[0], ball[1], 0, "");
        distance[ball[0]][ball[1]] = 0;
        vector<int> dx = {-1, 1, 0, 0};
        vector<int> dy = {0, 0, -1, 1};
        vector<char> direction = {'u', 'd', 'l', 'r'};
        while (!Q.empty()) {
            int x = Q.top().row;
            int y = Q.top().col;
            int d = Q.top().dist;
            string dir = Q.top().dir;
            Q.pop();
            if (x == hole[0] && y == hole[1]) {
                return dir;
            }
            
            for (int i = 0; i < dx.size(); ++i) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                int nd = d + 1;
                while (nx >= 0 && nx < m && ny >= 0 && ny < n && maze[nx][ny] == 0 && !(nx == hole[0] && ny == hole[1])) {
                    nx += dx[i];
                    ny += dy[i];
                    ++nd;
                }
                
                if (!(nx == hole[0] && ny == hole[1])) {
                    nx -= dx[i];
                    ny -= dy[i];
                    --nd;
                }
                
                if (distance[nx][ny] >= nd && !(nx == x && ny == y)) {
                    Q.emplace(nx, ny, nd, dir + direction[i]);
                    distance[nx][ny] = nd;
                }
            }
        }
        
        return "impossible";
    }
    
private:
    struct Cell {
        int row;
        int col;
        int dist;
        string dir;
        Cell(int r, int c, int d, string di): row(r), col(c), dist(d), dir(di){}
        bool operator < (const Cell& c) const {
            if (dist == c.dist) {
                return dir > c.dir;
            }
            return dist > c.dist;
        }
    };
};

File: 0401-0500/0499-The Maze III/0499-The Maze III.py
import heapq
import math
# Dijkstra
class Solution:
    def findShortestWay(self, maze: List[List[int]], ball: List[int], hole: List[int]) -> str:
        m, n = len(maze), len(maze[0])
        pq = [[0, ''] + ball]
        distance = [[math.inf] * n for _ in range(m)]
        distance[ball[0]][ball[1]] = 0
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]
        direction = ['u', 'd', 'l', 'r']
        while pq:
            d, di, r, c = heapq.heappop(pq)
            if [r, c] == hole:
                return di
            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]
                nd = d + 1
                while 0 <= nr < m and 0 <= nc < n and maze[nr][nc] == 0 and not [nr, nc] == hole:
                    nr += dr[i]
                    nc += dc[i]
                    nd += 1
                if not [nr, nc] == hole: 
                    nr -= dr[i]
                    nc -= dc[i]
                    nd -= 1

                if distance[nr][nc] >= nd and not [nr, nc] == [r, c]:
                    heapq.heappush(pq, [nd, di + direction[i], nr, nc])
                    distance[nr][nc] = nd
        
        return 'impossible'


# A*
class Solution:
    def findShortestWay(self, maze: List[List[int]], ball: List[int], hole: List[int]) -> str:
        m, n = len(maze), len(maze[0])
        pq = [[abs(hole[0] - ball[0]) + abs(hole[1] - ball[1]), 0, ''] + ball]
        distance = [[math.inf] * n for _ in range(m)]
        distance[ball[0]][ball[1]] = 0
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]
        direction = ['u', 'd', 'l', 'r']
        while pq:
            _, d, di, r, c = heapq.heappop(pq)
            if [r, c] == hole:
                return di
            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]
                nd = d + 1
                while 0 <= nr < m and 0 <= nc < n and maze[nr][nc] == 0 and not [nr, nc] == hole:
                    nr += dr[i]
                    nc += dc[i]
                    nd += 1
                if not [nr, nc] == hole: 
                    nr -= dr[i]
                    nc -= dc[i]
                    nd -= 1

                if distance[nr][nc] >= nd and not [nr, nc] == [r, c]:
                    heapq.heappush(pq, [nd + abs(hole[0] - nr) + abs(hole[1] - nc), nd, di + direction[i], nr, nc])
                    distance[nr][nc] = nd
        
        return 'impossible'
